### DISCUSSION TITLE
Request for Comments: `.xoz` new format - Object Specification

### DISCUSSION THREAD

**Discussion by eldipa** [xournalpp/xournalpp#4896] [Ideas]
Hi, I was working these days on a specification/proposal for the new `.xoz` file format that would replace the current `.xopp` file.

There are still a few open questions but I think it is reasonable for a "request for comments" round:
[RFC Object Spec](https://github.com/eldipa/xoz/wiki/RFC-Object-Spec)

I would like to get some feedback from the community and developers. 

I left a quick overview of the main concepts behind `.xoz` for those who want the TLDR version:

![image](https://user-images.githubusercontent.com/2665522/241585870-7fee7e8c-220d-48d9-8b3c-4aa5eb5d066b.png)

**Comment by tmoerschell** [comment]
Hi, I'm surely a bit late here. This is tremendous work that you have been doing. I hope that you continue to find the energy to move forward with this large project, and that we may one day see the results of all this work!

I would have a few comments, and I hope it is not much too late for these reflections to be integrated into your work. 

### General approach and integration into the Xournal++ application:

I think it would be beneficial to separate this project into 2 parts: 
 1. The random access file format: a library that is, given a description of the contained objects, capable of managing a file of arbitrary objects and streams. It manages the data blocks, keeps an index of the objects it read, and does the low level reading / writing of serialized data. 
 2. The binary serialization of Xournal++ objects: does the translation between Xournal++ objects and serialized data. The random access parser also relies on it to navigate the file properly. 

In my opinion, this would improve the maintainability of the project, as the code can be split into separate parts with a clear interface, instead of being a monolithic parser. The first part could be entirely agnostic of the actual data structure, and in theory even be used by applications other than Xournal++. 

Although the code isn't cleanly separated, the current file formats also work this way. The Xournal++ objects are serialized as XML and then contained in either a gzip or zip file (the zip file may also contains raw attachments). I did a small sketch to illustrate it a bit: 

![image](https://github.com/xournalpp/xournalpp/assets/144177788/b0715858-1d5d-4900-aecd-7030addc260b)

The existing load/save handler could be rewritten to allow a more modular approach, which will help with easily integrating a new file format. 

### Layers and pages:

I do not think that a layers-first approach would integrate well into Xournal++. Currently, each page can have a different set of layers. Additional layers are created only on a single page. Merging layers from different pages together could be a challenge when converting old files into new ones, as it may result in illogical combinations of layers. A layer-first format would also call for the in-application Xournal++ document structure to be changed, which would mean old documents need to be adapted on opening with a newer version. 

I see little use for modification or reordering of a layer across all pages. If there should be a need for this, such features could be implemented with the current document structure and file format, while keeping the possibility to make modifications on a single page layer. 

A user is much more likely to change elements on an individual page or append new pages, rather than doing modifications on a single layer. Considering large files, in a page-first approach, most pages will not even need to be read, and even less written to. On the other hand, in a layer-first approach, each layer has to be queried for a tile (you did mention this as a cheap operation), and each layer will be modified when a page is added, which is probably much more frequent than changes to layers across all pages. The page-first approach takes advantage of reading and writing being more local, closer together. While this is not a huge deal for a random access file format, it might still make a small difference. 

If the Xournal++ document structure is not changed, having a different representation in the file format will lead to more complexity and confusion between different meanings of the same names. 

### The mosaic:

There already is some support of page size modification, and it may be further developed with the current file format. I do however think that splitting up tiles (or single page layers) into multiple parts can be of interest. Not all elements of a tile may need to be loaded at once, especially considering the eventuality of an infinite page option. The application may also want to prioritize the loading of certain regions. It will be the duty of the application to implement an efficient splitting of the tiles, but hints from the library, such as knowing the remaining space in a tiles's stream or allocation might be useful. 



Finally, if I some of my explanations weren't clear enough, I will be happy to develop them further, and I would of course also welcome further discussion. Thank you for everything you have already done. Keep up the good work!

**Comment by eldipa** [comment]
Hi @tmoerschell , it is never too late. A matter of fact I was little active in the last few months to work on other projects and yesterday I went back to `xoz` so your message and support was at the perfect time. Thank you!

I'm going to answer your comments:

## General approach and integration into the Xournal++ application:

> I think it would be beneficial to separate this project into 2 parts:
>   1. The random access file format: a library that is [...]
>   2. The binary serialization of Xournal++ objects: does the translation between Xournal++ objects and serialized data. The random access parser also relies on it to navigate the file properly.

Totally agree. In fact I call these "milestones":

- Milestone 1: the library (as your step 1)
- Milestone 2: the serialization of the objects and the integration with `Xournal++` for load/save `xoz` files in sequential order (part of your step 2)
- Milestone 3: the modification of `Xournal++` to take advantage of the random access for loading/saving (part of your step 2)

Why split the step 2 into two milestones? While the `xoz` library support random access, `Xournal++` doesn't. The `LoadHandler` loads the entire `xopp` file and creates all the objects in memory in one call; `SaveHandler` scans all the objects in memory and writes the `xopp` file in sequential order.

Changing  `LoadHandler` and `SaveHandler` to load/save from `xoz` in sequential order (milestone 2) is totally possible. As you mentioned in your diagram, we can make those two classes to support `xoz`, `xopp`,  `xopz` and others. Milestone 2 in this regards is mostly refactor the code in those two classes and a little of adaptation.

Milestone 3 is another story. `Xournal++` assumes that the whole document and objects are in memory (so no reads from disk) and when it does a modification, it does not assume that it will have an impact on disk. It is unlikely that we can change that. My idea is to keep the illusion that everything is in memory while in fact it is read on demand. The same goes for writing: `Xournal++` can do any modification to the objects and the writes to disk are transparent (possibly on memory pressure or time).

Milestone 3 requires a much deeper understanding of how `Xournal++` manages the document, how it is the memory management and the concurrency. It is a quite large project in its own. The good notice is that we can ship `xoz` to production without it. It will not give us the best performance, but it will give us some improvements and the feedback of how `xoz` works in the real world.

> In my opinion, this would improve the maintainability of the project, as the code can be split into separate parts with a clear interface, [...]  in theory [the lib] even be used by applications other than Xournal++.

That's true. While I designed `xoz` thinking in `Xournal++`, the lib is quite flexible. It is not a general purpose but other applications can benefit from it. My intention is to have the agnostic code in separated github project and the `Xournal++`-specifics in the `xournalpp` project.

## Layers and pages:

> I do not think that a layers-first approach would integrate well into Xournal++. 

Me neither. The design document is totally unclear on this and badly written. It puts the focus on the `layer` while the basic unit of information is the `tile`. A `tile` has a coordinate `x,y,layer` (or `row,column,layer`). The `layer` in only a meaningful way to think in how the tiles are organized in a 2-d space: a `mosaic`.

Moreover, `xoz` will not store empty tiles. So, `xoz` models a document as a 3-d (`x,y,layer`) *sparse matrix*. (note: this is not written in stone, but this is going to be the general idea).

> Currently, each page can have a different set of layers. Additional layers are created only on a single page.

I thought that if I add a layer in `Xournal++` that would add the layer to every single page in the document but I'm may be wrong. In any case adding "a new layer" in `xoz` does not change much, only when the user draw something, some of the tiles of the layer will be actually stored. 

> Merging layers from different pages [...]

Could you elaborate more on this? I agree with you that merging layers of different pages (or  coordinates ) is of little use. 

> [The] document structure to be changed, which would mean old documents need to be adapted on opening with a newer version.

This is true by the mere introduction of `xoz` but I want to separate this in three "kind" of documents
 - the document stored in `xoz`, this is represented in terms of `tiles` as discussed above
 - the document "in memory" that `Xournal++` uses for rendering. For Milestone 1 and 2 this is untouched, for Milestone 3 this representation may change (discussed a little above)
 - the document that the plugins see, this should not change even in Milestone 3 so we keep compatibility.

When a future `Xournal++` (with the Milestone 3 integrated) reads a `xopp` file, it loads the file as usual and does any transformation required to make it work. This is transparent for the user and expected for us (the developers) to make it work. On save, if the user wants, it can store the file in `xoz`. Of course, new features in `Xournal++` and performance improvements will happen only with `xoz`.

> I see little use for modification or reordering of a layer across all pages. [...]
True, `xoz` does not impose any restriction on this. If the application wants to modify all the tiles of a layer, it can do it just iterating over each tile.

>[...] In a layer-first approach, each layer has to be queried for a tile (you did mention this as a cheap operation), and each layer will be modified when a page is added, which is probably much more frequent than changes to layers across all pages. The page-first approach takes advantage of reading and writing being more local, closer together.

Yes but no :D. If a new layer is added, by definition no drawing exists in that layer (yet) so nothing is written to `xoz`. If the user writes in one of the tiles of the layer, only that tile is written. *"You pay as you go"*.

In terms of what is the cost to render a page, you correctly guessed that you need to find and get all the tiles located in the position `x,y` of the page across all the layers. It looks expensive but it isn't. The `xoz` library will have in memory an index to all the tiles so it can get easily the requested ones. Then, it will do a fetch/read for each. This is probably some expensive because `xoz` does not guarantee any locality so in principle, it will end up reading `n` different locations in the file.

My hypothesis is that this is not going to hurt the performance because most of the `xopp` docs out there have 1 layer anyways (some have 2 o 3).
I'm not discarding to do some improvements in the locality, but for now, I'm leaving this for a next generation of `xoz`. My bet is that we can improve the locality changing the algorithm of the allocator without changing the file format so any next generation will be forward and backward compatible.

> If the Xournal++ document structure is not changed, having a different representation in the file format will lead to more complexity and confusion between different meanings of the same names.

Agree. I think I should make a document explicitly saying the mapping between `Xournal++` and `xoz` terminology, probably in the Milestone 2 or 3 but I would not like to bind how `Xournal++` models a document and how `xoz` does it because we may get less freedom. At the end, it is ok to have one way to store data in disk, one way to store data in memory and other way to present this information to the user.

> There already is some support of page size modification, [...] splitting up tiles (or single page layers) into multiple parts can be of interest.

For `xoz` perspective, the tiles does not have a size in terms of pixels or cm, so if a page changes its size, it is represented by the same tile (more specifically, a stack of tiles from all the layers, as it is in the diagram above). The idea to have more tiles to the sides is to support writing things "in the margin" and more generally to have an "infinite canvas". 
It is up to `Xournal++` to do that: it may decide to store absolutely everything in the infinite canvas in a single tile or (more wisely), split the infinite canvas into a mosaic and use multiples tiles.. `xoz` does not enforce either but will have support for both.

Well,  @tmoerschell , it was a long message! but I think I covered all. Let me know if I missed something or something was not clear. It was a pleasure to write about this and I take some of your comments for consideration, even if it is not for the version 1 of `xoz` but for a version 2.

Thanks a lot for the interest!
Martin.

**Reply by tmoerschell** [reply]
In Xournal++, layers are entirely owned by a page. You can add, remove, move and rename layers any way you want, other pages keep their own layers and will not be affected in any way.

With the current design, you can have on one page a background named "Background", and two layers named "Layer 1" and "Corrections", and on the next page, a background layer named "Not Background" and 3 layers named "My Layer 4", "One more Layer" and "Layer 1". If such a file is converted to the new file format, there is no way to differentiate which pages had which layers. If Xournal++ should merge the layers that are on the same level, how can the different names on different pages be kept? If it should preserve all layers, how would the application know that a particular tile doesn't belong to a page (or if it is just empty)? 

 > [Adaptation of old files] is true by the mere introduction of `xoz`

It does not need to be. I think we should aspire to be able to convert a .xopp file into a .xoz file and back again with no noticeable difference to the original. Adding functionality is good and expected, but removing or modifying existing features must be thoroughly considered.

I would propose the following general file structure: 
 * The document contains pages
 * Pages contain layers
 * Layers contain one or more tiles
 * Tiles contain elements (text, strokes, and so on)

 > I would not like to bind how Xournal++ models a document and how xoz does it because we may get less freedom. At the end, it is ok to have one way to store data in disk, one way to store data in memory and other way to present this information to the user.

It is ok to have differences if the new way makes more sense and allows significant improvements. However, I cannot see how the change in structure could help us, and I fear difficulties for the integration with the application. Having document models be aligned between disk, memory and user interface allows for the least complexity, which avoids errors and weird bugs.  


On another note, I have noticed that you did not mention attaching external PDFs (or other files) in the specification. Is this still in your thoughts? There were some comments in favor of this in the main issue, and I think it is a nice feature that should be kept. 

If possible, I would like to provide some help on this project. I think I would be starting to look at how things inside Xournal++ can be prepared for later "milestones", starting with a redesign of the Load/SaveHandler. 

Kindly,
Thomas

**Comment by eldipa** [comment]
> IYou can add, remove, move and rename layers any way you want, other pages keep their own layers and will not be affected in any way.

That's true, I didn't know that! The good thing is that `xoz` supports this out of the box.

> If such a file is converted to the new file format, there is no way to differentiate which pages had which layers.

There is no trouble here. Let's say you have 2 pages with backgrounds "Background" and "No Background".
These will be represented by 2 tiles: `(page 1,  z=0, "Background")` and `(page 2, z=0,  "No Background")`

Now, let's add some layers. As you mentioned a layer is a page-specific thing. We can represent them as tiles too:
 - `(page 1,z=1,  "Layer 1")`
 - `(page 1, z=2, "Corrections")`
 - `(page 2, z=1,  "My Layer 4")`
 - `(page 2, z=2, "One more Layer")`
 - `(page 2, z=3, "Layer 1)`

> If Xournal++ should merge the layers that are on the same level, how can the different names on different pages be kept? If it should preserve all layers, how would the application know that a particular tile doesn't belong to a page (or if it is just empty)?

There is no need to merge anything. If `Xournal++` wants to render the page 1, it grabs all the tiles for that page, sort them by `z` and renders them in order.

I understand that the format specification seems to point in another direction and the mention of the word "layer" is giving some misunderstanding. I will change the specification to make it clear.

By the way, in reality, in `(page 1,z=1, ...)` the "page 1" is encoded as 2-d coordinate so it is `(x=0, y=1, z=1, ...)`. This enable us to create as many pages in all the directions, more like an infinite canvas.

> > [Adaptation of old files] is true by the mere introduction of `xoz` 
> It does not need to be. I think we should aspire to be able to convert a .xopp file into a .xoz file and back again with no noticeable difference to the original.

I agree but it does not imply that `xoz` should be similar to `xopp`. Is like storing an image in `jpeg` and `bmp` formats. The image looks the same but the file formats are very different.

> Adding functionality is good and expected, but removing or modifying existing features must be thoroughly considered.
Absolutely agree. No feature should be removed. `xoz` aims to be efficient and flexible enough to support current and future features.

> I would propose the following general file structure [....]

With the coordinates `(x, y, z)` the file structure become flat:
- a document has tiles
- a tile has a coordinate `(x, y, z)` (and possibly a name) and contains elements (strokes, text, ...)


> It is ok to have differences [between how we store in disk and how we store in memory] if the new way makes more sense and allows significant improvements. However, I cannot see how the change in structure could help us, and I fear difficulties for the integration with the application

In the motivation section I wrote the limitations of the current `xopp` in terms of performance and why a binary solution is required. This is a pretty good incentive to do the change.

> Having document models be aligned between disk, memory and user interface allows for the least complexity, which avoids errors and weird bugs.

True but there are tradeoff. Back to the `jpeg` / `bmp` example, we the your reasoning we should be using `bmp` as it fits more nicely on how the computers work. But there is a catch. `jpeg` has much more compression ratio, it was designed to work better with our eyes and not with the computers.
It is more complex parsing/writing a `jpeg` than `bmp`? Absolutely. No doubts on that. Is more likely to have bugs? Hell yes! But again, there are tradeoffs.

With `xopp` and `xoz` is the same. 

> On another note, I have noticed that you did not mention attaching external PDFs (or other files) in the specification. Is this still in your thoughts? There were some comments in favor of this in the main issue, and I think it is a nice feature that should be kept.

Absolutely yes. And some more things that I found in other open issues.

> If possible, I would like to provide some help on this project. I think I would be starting to look at how things inside Xournal++ can be prepared for later "milestones", starting with a redesign of the Load/SaveHandler.

A refactor to `Load/SaveHandler` is truly needed. Right now it does all the `xopp`/`xoj`/`xopz` parsing in the same spot.
